### PR TITLE
Sol database

### PR DESCRIPTION
**Pull request by Sosol02**

### CODE REVIEW

File: app/src/main/java/com/github/onedirection/database/ConcreteDatabase.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+import com.google.firebase.firestore.DocumentSnapshot;
+import com.google.firebase.firestore.FirebaseFirestore;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * The auto synced database of the application
+ */
+public class ConcreteDatabase /* implements Database */ {
```
Comment (Ef55) [inline app/src/main/java/com/github/onedirection/database/ConcreteDatabase.java:18]: This should be split into _interface ConcreteDatabase_ and _class ConcreteFirebaseDatabase extends ConcreteDatabse_

File: app/src/main/java/com/github/onedirection/database/ConcreteDatabase.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    private final FirebaseFirestore db;
+
+    private static final ConcreteDatabase global = new ConcreteDatabase();
+
+    public static ConcreteDatabase getDatabase() {
```
Comment (Ef55) [inline app/src/main/java/com/github/onedirection/database/ConcreteDatabase.java:23]: The name should be made coherent with the naming scheme used in the authentication module (either by changing this class or AuthenticationService).

File: app/src/main/java/com/github/onedirection/database/store/Id.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
+package com.github.onedirection.database.store;
+
+import java.util.Objects;
+import java.util.UUID;
+
+public class Id {
+    // these fields are what's compared for equality
+    // the UUID are extremely unique (2^122 possibilities)
```
Comment (Ef55) [inline app/src/main/java/com/github/onedirection/database/store/Id.java:8]: Yes, much uniqueness.

File: app/src/main/java/com/github/onedirection/database/store/Storer.java [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package com.github.onedirection.database.store;
+
+/**
+ * Any object that wants to be stored in the database needs
+ * to implement it.
+ * @T The type being stored
+ */
+public abstract class Storer<T extends Storable<T>> {
+    public abstract DatabaseCollection getCollection();
+    public abstract Class<T> classTag();
```
Comment (Ef55) [inline app/src/main/java/com/github/onedirection/database/store/Storer.java:10]: I think this should be documented.

File: app/src/test/java/com/github/onedirection/IdUnitTest.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+import com.github.onedirection.database.store.Id;
+
+import org.junit.Test;
+
+import java.util.ArrayList;
+
+import static org.junit.Assert.*;
+
+public class IdUnitTest {
+    private static final int testNum = 500;
```
Comment (Ef55) [inline app/src/test/java/com/github/onedirection/IdUnitTest.java:12]: This should be SCREAMING_SNAKE_CASE :)

**Review by Ef55 — APPROVED** [review]

File: app/src/main/java/com/github/onedirection/database/ConcreteDatabase.java (hunk 1 above)
Comment (Sosol02) [reply app/src/main/java/com/github/onedirection/database/ConcreteDatabase.java:18]: Plutôt interface Database et class FirebaseDatabase

**Review by Flechman — APPROVED** [review]

**Pull request merged**